src/text/non_blocking/text_buffer.rs

```diff
@@ -284,6 +284,7 @@ impl TextBuffer<Vec<u8>> {
 
         let read_buffer = &mut self.data.as_mut_slice()[self.data_end..];
         let bytes_read = source.read(read_buffer)?;
+        self.data.resize(self.data_end + bytes_read, 0);
```

Comment:
oof.. Thank you for calling that out. I had run into an issue where parsing was failing due to unexpected characters in the text (large number of '\0's), that was ultimately due to the `read_from` reserving the needed space for the requested bytes. I quickly fixed it with this resize, to truncate the buffer down to only what was read.  At the time I was focused on another area of the code, and never went back to investigate why the issue was happening.

I've dug deeper now, and I'll post a new commit that removes this resize in favor of fixing the underlying bug in the text buffer. 

When UTF-8 validation is done, my original code validated `&self.data.as_ref()[valid_end..]`. Once a `read_from` happens that pre-extends the buffer beyond the actual data size the UTF-8 validation starts validating the end of the buffer without read data. Since the data is initialized with '\0's, the unread bytes do not get evaluated by `from_utf8`, however the `Ok(_)` match assumed the entire buffer was valid and would return the buffer's length as the validated end.